### PR TITLE
[BasicSpecification] Dont consider extname in have_file?

### LINKED ISSUE #1507: Requiring an .so file via Ruby will not work with RubyGems
If I require a .so file via Ruby's require, it works.  But if I do the same thing via RubyGems, it won't work.  Here is an example:

```
# Put nokogiri's lib dir on the load path so that RubyGems isn't used, then require the .so file
[aaron@TC ~]$ ruby -I $(dirname `gem which nokogiri`) -e'l = $LOADED_FEATURES.dup; require "nokogiri/nokogiri.so"; p $LOADED_FEATURES - l'
["/Users/aaron/.rbenv/versions/ruby-trunk/lib/ruby/gems/2.4.0/gems/nokogiri-1.6.8.rc3-x86_64-darwin-15/lib/nokogiri/nokogiri.bundle"]
```

Ruby automatically deals with the .so -> .bundle conversion (so that the require is cross platform).

If I remove Nokogiri from the load path, then RubyGems will not behave the same way Ruby does:

```
[aaron@TC ~]$ ruby -e'l = $LOADED_FEATURES.dup; require "nokogiri/nokogiri.so"; p $LOADED_FEATURES - l'
/Users/aaron/.rbenv/versions/ruby-trunk/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- nokogiri/nokogiri.so (LoadError)
    from /Users/aaron/.rbenv/versions/ruby-trunk/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    from -e:1:in `<main>'
[aaron@TC ~]$
```

I expect the `require` call to work without being changed, but what actually happens is that RubyGems raises an exception.

### PR DESCRIPTION
**Pull request by segiddins**
# Description:

An attempt to fix #1507. Still needs test coverage.

\c @tenderlove 
# Tasks:
- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends
- [x] [Squash commits](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html)

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by tenderlove** [comment]
I think this will result in false positives.  For example, consider a gem that contains *only* `foo.rb`, but the user code says `require "foo.so"`.  This will return true even though the actual require will fail.

I *think* the best way to fix this problem is to sub '.so' with DLEXT for the current platform.

**Comment by indirect** [comment]
Can we just copy the MRI implementation inside `require`?

**Comment by segiddins** [comment]
@indirect I think at the very least we need to use MRI's `require` test suite :/

**Comment by segiddins** [comment]
@tenderlove I think this does what you want now?

File: lib/rubygems/util.rb [hunk 1]
```diff
@@ -132,4 +132,14 @@ def self.traverse_parents directory
     end
   end
 
+  def self.sub_ext(file, extension)
+    basename = file.to_s[0..(-1 - File.extname(file))]
```
Comment (djberg96) [inline lib/rubygems/util.rb]: Don't think this works:
```
irb(main):001:0> file = "foo.so"
=> "foo.so"
irb(main):002:0> basename = file.to_s[0..(-1 - File.extname(file))]
TypeError: String can't be coerced into Fixnum
	from (irb):2:in `-'
	from (irb):2
	from /Users/dberger/.rbenv/versions/2.2.3/bin/irb:11:in `<main>'
```
What not just: ```File.join(File.dirname(file), File.basename(file, File.extname(file)))```

**Comment by segiddins** [comment]
@tenderlove ping to review this?

**Comment by segiddins** [comment]
Since we never got confirmation this fixes the reported issue, I'm going to presume we don't need to merge this?

**Comment by indirect** [comment]
@segiddins Bundler fixes this issue by un-monkeypatching require—it's a real bug in RubyGems, though. The example in the original issue works to demonstrate that this PR causes RubyGems to raise a different exception:

```
~/src/rubygems/rubygems (pr/1521) $ ruby -I lib -r rubygems -e'l = $LOADED_FEATURES.dup; require "nokogiri/nokogiri.so"; p $LOADED_FEATURES - l'
/Users/andre/src/rubygems/rubygems/lib/rubygems/basic_specification.rb:316:in `have_file?': uninitialized constant Gem::Util (NameError)
	from /Users/andre/src/rubygems/rubygems/lib/rubygems/basic_specification.rb:79:in `contains_requirable_file?'
	from /Users/andre/src/rubygems/rubygems/lib/rubygems/specification.rb:1037:in `block in find_by_path'
	from /Users/andre/src/rubygems/rubygems/lib/rubygems/specification.rb:1036:in `each'
	from /Users/andre/src/rubygems/rubygems/lib/rubygems/specification.rb:1036:in `find'
	from /Users/andre/src/rubygems/rubygems/lib/rubygems/specification.rb:1036:in `find_by_path'
	from /Users/andre/src/rubygems/rubygems/lib/rubygems.rb:189:in `try_activate'
	from /Users/andre/src/rubygems/rubygems/lib/rubygems/core_ext/kernel_require.rb:126:in `rescue in require'
	from /Users/andre/src/rubygems/rubygems/lib/rubygems/core_ext/kernel_require.rb:40:in `require'
	from -e:1:in `<main>'
```

With nokogiri installed but without using RubyGems to do the require, it works:
```
$ ruby -I $(dirname `gem which nokogiri`) -e'l = $LOADED_FEATURES.dup; require "nokogiri/nokogiri.so"; p $LOADED_FEATURES - l'
["/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/gems/2.0.0/gems/nokogiri-1.5.6/lib/nokogiri/nokogiri.bundle"]

With just RubyGems, it doesn't work:
```
$ ruby -e'l = $LOADED_FEATURES.dup; require "nokogiri/nokogiri.so"; p $LOADED_FEATURES - l'l
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- nokogiri/nokogiri.so (LoadError)
	from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from -e:1:in `<main>'
```

With Bundler, require is back to the default, so it works: 
```
$ bundle exec ruby -e'l = $LOADED_FEATURES.dup; require "nokogiri/nokogiri.so"; p $LOADED_FEATURES - l'
["/Users/andre/src/rubygems/rubygems.org/.bundle/ruby/2.3.0/gems/nokogiri-1.6.7.2/lib/nokogiri/nokogiri.bundle"]
```

**Comment by segiddins** [comment]
@indirect in that case, does this patch fix things?

**Comment by indirect** [comment]
@segiddins no, this patch causes the first exception in my comment (and demonstrates a command you can use to test the patch yourself)

**Comment by segiddins** [comment]
@indirect thanks for the pointers -- fixed and added a test

**Comment by duckinator** [comment]
So, what's the status of this? It looks like it has failing tests and probably needs to be rebased (because it's a few months old)?

**Comment by segiddins** [comment]
If I recall, was waiting on review from Andre or Aaron

**Comment by indirect** [comment]
@segiddins can we put the method inside `basic_specification.rb`, since that's the only place it gets used? I want to be really, really conservative about how much code gets loaded by `rubygems.rb`, and if we don't need the rest of `util.rb`, let's skip it.

**Comment by segiddins** [comment]
@indirect done

**Comment by indirect** [comment]
@segiddins I am 👍 on this once the tests pass!

**Comment by segiddins** [comment]
Failures appear to be ruby version related :(

**Comment by deivid-rodriguez** [comment]
@bronzdoc Is this no longer an issue? I prefer PRs that we could potentially revive and make use of to stay open.

**Pull request closed without merging**